File: EDA_scripts/tb_auto_gen.py
import sys
import re

module_name_ori = sys.argv[1]
module_name = re.sub(r'.v', '', module_name_ori)

fout = open('tb_'+module_name_ori,'wt') 
fout.write('`timescale 1ns/1ps\n')
fout.write('module ' + module_name + ';' + '\n')

# get all input and output ports
fin = open(module_name_ori,'rt')
in_ports_no_comma_list = []
out_ports_no_comma_list = []
for line in fin:
    # match input in1,in2, or input [3:0] in1,in2,
    m = re.match(r'^\s*input\s*(?P<WIDTH>(\[\d*:\d\])*)\s*(?P<NAME>(\w+\,\s*)+)',line) 
    if m:
        in_ports = m.group('NAME')
        port_width = m.group('WIDTH')
        in_ports_no_comma_tmp = re.split(',',in_ports)
        in_ports_no_comma_list += in_ports_no_comma_tmp[0:-1]

    # match output in1,in2, or input [3:0] in1,in2, pr output out3
    m1 = re.match(r'^\s*output\s*(?P<WIDTH>(\[\d*:\d\])*)\s*(?P<NAME>(\w+\,\s*)+\w+)',line) 
    if m1:
        out_ports = m1.group('NAME')
        port_width = m1.group('WIDTH')
        out_ports_no_comma_tmp = re.split(',', out_ports)
        out_ports_no_comma_list += out_ports_no_comma_tmp

for ports in in_ports_no_comma_list:
    if port_width:
        fout.write('reg ' + port_width + ' ' + ports + ';\n')
    else:
        fout.write('reg' + port_width + ' ' + ports + ';\n')

for ports in out_ports_no_comma_list:
    if port_width:
        fout.write('wire ' + port_width + ' ' + ports + ';\n')
    else:
        fout.write('wire' + port_width + ' ' + ports + ';\n')

# instance module
fout.write('\n//instance\n')
fout.write(module_name + ' ' + module_name + '_i(\n')
for ports in in_ports_no_comma_list:
    fout.write('\t.' + ports + '(' + ports + '),\n')
for ports in out_ports_no_comma_list[0:-1]:
    fout.write('\t.' + ports + '(' + ports + '),\n')
#fout.write('\t.' + out_ports_no_comma_list[-1] + '(' + out_ports_no_comma_list[-1] + ')\n')
fout.write(');\n\n')

for ports in out_ports_no_comma_list:
    fout.write(ports)

# get clock ports
clk_ports = {} # dict store all clk ports
for ports in in_ports_no_comma_list:
    m2 = re.match(r'\w*rst\w*', ports)
    m3 = re.match(r'\w*clk\w*', ports)
    if m2:
        rst_port = m2.group()
    if m3:
        clk_ports[m3.group()] = ''

num_str = '01234'
num = 0
for ports in clk_ports.keys():
    fout.write('localparam CLK_PERIOD_' + num_str[num] + ' = 10;\n')
    fout.write('always #(CLK_PERIOD_' + num_str[num] + '/2.0) ' + ports + '= ~'+ ports + '\n')
    fout.write('\n')
    num += 1

fout.write('initial begin\n')
for ports in in_ports_no_comma_list:
    fout.write('\t' + ports + ' = 0;\n')
fout.write('\t//removal reset after 50 ns\n')
fout.write('\t' +  rst_port + ' = 1;\n')
for clk in clk_ports.keys():
    fout.write('\t@(posedge ' + clk + ');\n')
fout.write('\t//add stimu here\n\n\n\n')
fout.write('end\n\n')

fout.write('initial begin\n')
fout.write('\t$fsdbDumpfile(\"' + module_name + '.fsdb\");\n')
fout.write('\t$fsdbDumpvars(0,tb_' + module_name + ');\n')
fout.write('\t$fsdbDumpMDA();\n')
fout.write('end\n\n')
fout.write('endmodule\n\n')

File: EDA_scripts/perl/tb_auto_gen.pl
#!/usr/bin/perl -w 
use 5.010;
  
say "*"x30;
say "function: generating testbench file automaticly";
say "*"x30;

my @all_in_var; # store all input variable
my @all_out_var; # store all input variable
my @all_var;  # store all variable in the module

#===============================#
# get argument from command line
#===============================#
my $in = $ARGV[0];
if(! defined $in){
  die "usage:$0 filename";
}

# output file name
my $out = $in;
$out =~ s/^(w+)?/tb_/g;

# get module name
my $module_name = $in;
$module_name =~ s/\.v//g;

# open file
if (!open $in_fh,'<',$in){
  die "can't open '$in': $!";
}

if (!open $out_fh, '>',$out){
  die "can't write '$out': $!";
}

select $out_fh;
say "`timescale 1ns/1ps\n";
say "module tb_$module_name;";
select $in_fh;

#=====================================
# find input and output in the module
#=====================================
while (<$in_fh>){
  select $out_fh;
  chomp;
  # patten input or output
  if(/\b(in|out)put\b/){
    #say  $_;
    my $tmp = $';
    # input 
    if("$&" eq "input"){
      $tmp =~ s/\,$//g; 
      # patten multi variable at one line or just one
      if($tmp =~ m/(((\w+\s?\,\s?)+)?\w+\b$)/){
        my $string = $&;
        my @mult_var = split /,/, $string;
        # remove all space
        foreach $var_tmp (@mult_var){
          $var_tmp =~ s/(\s?\b|\b\s?)//g;
          push @all_in_var, $var_tmp;
        }
      }
      say "reg$tmp;";    # eg: reg [3:0] a;
    }
    # output 
    if("$&" eq "output"){
      $tmp =~ s/\,$//g;   # remove the last comma
      # remove reg if the output in the module is reg type
      $tmp =~ s/\breg\s?\b//g; 
      # patten multi variable at one line or just one
      if($tmp =~ m/(((\w+\s?\,\s?)+)?\w+\b$)/){
        my $string1 = $&;
        my @mult_var1 = split /,/, $string1;
        # remove all space
        foreach $var_tmp1 (@mult_var1){
          $var_tmp1 =~ s/(\s?\b|\b\s?)//g;
          push @all_out_var, $var_tmp1;
        }
      }
      say "wire$tmp;";   #eg: wire [3:0] b;
    }
  }
}

select $out_fh;

#===================================#
# instance module in the testbench
#===================================#
push @all_var,@all_in_var;
push @all_var,@all_out_var;
say "\n// instance";
print "$module_name $module_name";
print "_inst(\n";
$var_num = @all_var - 1;
foreach (0..$var_num){
  if ($_ == $var_num){
    say "\t\.$all_var[$_]($all_var[$_])";
  }
  else{
    say "\t\.$all_var[$_]($all_var[$_]),";
  }
}
say ");";

#====================================#
# find all reset and all clk
#====================================#
foreach (@all_in_var){
  if(/(clk|clock)/){
    push @clk,$_;
  }
  if(/(rst|reset)/){
    push @rst,$_;
  }
}

say "\n\/\/generate all clk";
$clk_num = @clk-1;
foreach (0..$clk_num){
  say "localparam CYCLE_TIME_$_ = 10;";
  say "always #(CYCLE_TIME_$_/2.0)  $clk[$_] = ~$clk[$_];\n";
}

#==============================#
#add inital block in tb
#==============================#
say "\ninitial begin";
foreach (@all_in_var){
  say "\t$_ = 0;";
}
say "\t\/\/removal reset after 50 ns";
say "\t\#50;";
foreach (@rst){
  say "\t$_ = 1;";
}
say "\t\/\/add stimu here";
say "\n\n\n\n\n";
say "\t\#1000 \$finish;";
say "end";

#================================#
#add dump function in tb
#================================#
say "\ninitial begin";
say "\t\$fsdbDumpfile(\"$module_name.fsdb\");";
say "\t\$fsdbDumpvars(0,tb_$module_name);";
say "\t\$fsdbDumpMDA();";
say "end";

say "\nendmodule\n"

